File: src/Befunge93Strategy.java
/**
 * @file Befunge93Strategy.java
 * Strategy for Befunge-93 commands.
 * @author Conlan Wesson
 */

class Befunge93Strategy extends Unefunge93Strategy{
	public Befunge93Strategy(Field f, InstructionPointer i, StackStack s){
		super(f, i, s);
	}
	
	public boolean execute(char cmd){
		switch(cmd){
			case '^':
				ip.setDir(0, -1); break;
			case 'v':
				ip.setDir(0, 1); break;
			case '|':
				if(stack.top().pop() == 0){
					ip.setDir(0, 1);
				}else{
					ip.setDir(0, -1);
				}
				break;
			case '?':{
				int r = rnd.nextInt(4);
				switch(r){
					case 0:
						ip.setDir(0, 1); break;
					case 1:
						ip.setDir(-1, 0); break;
					case 2:
						ip.setDir(1, 0); break;
					case 3:
						ip.setDir(0, -1); break;
				}
				break;
			}
			default:
				return super.execute(cmd);
		}
		return true;
	}
}

File: src/Befunge93Strategy.cpp
/**
 * @file Befunge93Strategy.cpp
 * Strategy for Befunge-93 commands.
 * @author Conlan Wesson
 */

#include "Befunge93Strategy.h"
#include "FungeRunner.h"

namespace Funge {


Befunge93Strategy::Befunge93Strategy(FungeRunner& r) :
	FungeStrategy(r)
{
	r.pushSemantic('^', std::bind(&Befunge93Strategy::instructionNorth, this), FungeSemantic::MOVEMENT);
	r.pushSemantic('v', std::bind(&Befunge93Strategy::instructionSouth, this), FungeSemantic::MOVEMENT);
	r.pushSemantic('|', std::bind(&Befunge93Strategy::instructionIf, this));
}

FungeError Befunge93Strategy::instructionNorth(){
	ip.setDelta(Vector{0, -1});
	return ERROR_NONE;
}

FungeError Befunge93Strategy::instructionSouth(){
	ip.setDelta(Vector{0, 1});
	return ERROR_NONE;
}

FungeError Befunge93Strategy::instructionIf(){
	if(stack.top().pop() == 0){
		ip.setDelta(Vector{0, 1});
	}else{
		ip.setDelta(Vector{0, -1});
	}
	return ERROR_NONE;
}

FungeStrategy* Befunge93Strategy::clone(FungeRunner& r) const{
	return new Befunge93Strategy(r);
}

}
